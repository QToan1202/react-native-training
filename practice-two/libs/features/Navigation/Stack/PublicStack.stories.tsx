import { ComponentMeta, ComponentStory } from '@storybook/react-native'
import { NavigationContainer } from '@react-navigation/native'

import Stack from './index'

export default {
  title: 'navigation/Public Stack Navigation',
  component: Stack.PublicStackNavigator,
  decorators: [
    (Story) => (
      <NavigationContainer>
        <Story />
      </NavigationContainer>
    ),
  ],
} as ComponentMeta<typeof Stack.PublicStackNavigator>

const Template: ComponentStory<typeof Stack.PublicStackNavigator> = () => (
  <Stack.PublicStackNavigator />
)

export const Default = Template.bind({})
Default.args = {}
