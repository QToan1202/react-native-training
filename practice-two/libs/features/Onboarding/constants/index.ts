// eslint-disable-next-line import/prefer-default-export
export { default as ONBOARDING_VIEW_DATA } from './data'
