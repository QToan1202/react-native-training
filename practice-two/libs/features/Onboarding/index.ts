export * from './constants'
export { default as Onboarding } from './screens'
export * from './screens'
