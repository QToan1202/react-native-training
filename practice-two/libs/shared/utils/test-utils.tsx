import { render, RenderOptions } from '@testing-library/react-native'
import React, { ReactElement } from 'react'
import { NavigationContainer } from '@react-navigation/native'
import { SafeAreaProvider } from 'react-native-safe-area-context'

const AllTheProviders = ({ children }: { children: ReactElement }) => {
  return (
    <SafeAreaProvider>
      <NavigationContainer>{children}</NavigationContainer>
    </SafeAreaProvider>
  )
}

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react-native'

export { customRender as render }
