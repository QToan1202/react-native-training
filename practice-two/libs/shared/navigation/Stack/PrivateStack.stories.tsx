import { ComponentMeta, ComponentStory } from '@storybook/react-native'
import { NavigationContainer } from '@react-navigation/native'

import Stack from './index'

export default {
  title: 'navigation/Private Stack Navigation',
  component: Stack.PrivateStackNavigator,
  decorators: [
    (Story) => (
      <NavigationContainer>
        <Story />
      </NavigationContainer>
    ),
  ],
} as ComponentMeta<typeof Stack.PrivateStackNavigator>

const Template: ComponentStory<typeof Stack.PrivateStackNavigator> = () => (
  <Stack.PrivateStackNavigator />
)

export const Default = Template.bind({})
Default.args = {}
