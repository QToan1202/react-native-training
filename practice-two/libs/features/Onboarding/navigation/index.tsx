import { createNativeStackNavigator } from '@react-navigation/native-stack'

import { RootStackParamList } from '@practice-two/shared/types'
import { Feature } from '@practice-two/shared/hocs'
import { COLORS } from '@practice-two/shared/constants'

const Stack = createNativeStackNavigator<RootStackParamList>()

const OnboardingStack = () => (
  <Feature feat="onboarding">
    <Stack.Navigator screenOptions={{ headerShown: false, statusBarColor: COLORS.PRIMARY }}>
      <Stack.Screen name="Onboarding" getComponent={() => require('../screens').default} />
    </Stack.Navigator>
  </Feature>
)

export default OnboardingStack
