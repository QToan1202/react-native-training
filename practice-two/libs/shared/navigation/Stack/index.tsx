import { lazy } from 'react'
import { NativeStackHeaderProps, createNativeStackNavigator } from '@react-navigation/native-stack'
import { NavigatorScreenParams } from '@react-navigation/native'

import { COLORS } from '../../constants'
import BottomNav, { TabParamsList } from '../Tab'

// Components
const BackBar = lazy(() => import('../../components/Bar/Instance/Back'))
const CategoryBar = lazy(() => import('../../components/Bar/Instance/Category'))

type PublicStackParamsList = {
  Onboarding: undefined
  Login: undefined
  SignUp: undefined
}

type HomeStackParamsList = {
  Home: undefined
  CategoryDetail: { name: string }
  ProductDetail: { id: string } // ID of the product
  Cart: undefined
  AddAddress: undefined
  AddCard: undefined
  Payment: undefined
  OrderDetail: { id: string } // ID of the order
  Wishlist: undefined
}

type BrowseStackParamsList = {
  Browse: { search: string } | undefined
  ProductDetail: { id: string }
  Cart: undefined
  AddAddress: undefined
  AddCard: undefined
  Payment: undefined
  OrderDetail: { id: string } // ID of the order
  Wishlist: undefined
}

export type RootStackParamList = {
  Tabs: NavigatorScreenParams<TabParamsList>
  HomeStack: NavigatorScreenParams<HomeStackParamsList>
  BrowseStack: NavigatorScreenParams<BrowseStackParamsList>
  StoreStack: undefined
  OrderHistoryStack: undefined
  ProfileStack: undefined
} & PublicStackParamsList &
  HomeStackParamsList &
  BrowseStackParamsList &
  TabParamsList

const Stack = createNativeStackNavigator<RootStackParamList>()
const CustomHeader = (Element: React.JSX.ElementType, props: NativeStackHeaderProps) => (
  <Element {...props} />
)

const PublicStackNavigator = () => (
  <Stack.Navigator
    initialRouteName="Onboarding"
    screenOptions={{ headerShown: false, statusBarColor: COLORS.PRIMARY }}
  >
    <Stack.Screen
      name="Onboarding"
      getComponent={() => require('../../../features/Onboarding/screens').default}
    />
    <Stack.Screen
      name="Login"
      getComponent={() => require('../../../features/Login/screens').default}
    />
    <Stack.Screen
      name="SignUp"
      getComponent={() => require('../../../features/SignUp/screens').default}
      options={{
        headerShown: true,
        headerTitle: '',
        headerStyle: { backgroundColor: COLORS.PRIMARY },
        headerShadowVisible: false,
        headerTintColor: COLORS.WHITE,
      }}
    />
  </Stack.Navigator>
)

const TabBarStack = (
  <Stack.Group
    screenOptions={{
      header: (props: NativeStackHeaderProps) => CustomHeader(BackBar, props),
    }}
  >
    <Stack.Screen
      name="Wishlist"
      getComponent={() => require('../../../features/Wishlist/screens').default}
      options={{ headerTitle: 'wishlist' }}
    />
    <Stack.Screen
      name="Cart"
      getComponent={() => require('../../../features/Checkout/screens/Cart').default}
      options={{ headerTitle: 'my cart' }}
    />
    <Stack.Screen
      name="AddAddress"
      getComponent={() => require('../../../features/Checkout/screens/AddAddress').default}
      options={{ headerTitle: 'add a new address' }}
    />
    <Stack.Screen
      name="AddCard"
      getComponent={() => require('../../../features/Checkout/screens/AddCard').default}
      options={{ headerShown: false }}
    />
    <Stack.Screen
      name="Payment"
      getComponent={() => require('../../../features/Checkout/screens/Payment').default}
      options={{ headerTitle: 'payment option' }}
    />
    <Stack.Screen
      name="OrderDetail"
      getComponent={() => require('../../../features/Checkout/screens/OrderDetail').default}
      options={{ headerTitle: 'order details' }}
    />
  </Stack.Group>
)

const HomeStack = () => (
  <Stack.Navigator>
    <Stack.Screen
      name="ProductDetail"
      getComponent={() => require('../../../features/Product/screens').default}
      options={{ headerShown: false }}
    />
    <Stack.Screen
      name="CategoryDetail"
      getComponent={() => require('../../../features/Category/screens').default}
      options={{
        header: (props: NativeStackHeaderProps) => CustomHeader(CategoryBar, props),
      }}
    />
    {TabBarStack}
  </Stack.Navigator>
)

const BrowseStack = () => (
  <Stack.Navigator>
    <Stack.Screen
      name="ProductDetail"
      getComponent={() => require('../../../features/Product/screens').default}
      options={{ headerShown: false }}
    />
    {TabBarStack}
  </Stack.Navigator>
)

const PrivateStackNavigator = () => (
  <Stack.Navigator screenOptions={{ headerShown: false, statusBarColor: COLORS.PRIMARY }}>
    <Stack.Screen name="Tabs" component={BottomNav} />
    <Stack.Screen name="HomeStack" component={HomeStack} />
    <Stack.Screen name="BrowseStack" component={BrowseStack} />
  </Stack.Navigator>
)

export default { PrivateStackNavigator, PublicStackNavigator }
