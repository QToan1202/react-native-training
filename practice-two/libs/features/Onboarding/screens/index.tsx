import { useCallback, useEffect, useMemo, useRef, useState } from 'react'
import { Image, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import PagerView, { PagerViewOnPageSelectedEvent } from 'react-native-pager-view'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import { XStack } from 'tamagui'

import {
  Button,
  Paragraph,
  containerStyles,
  IOnboardingScreenView,
  Feature,
} from '@practice-two/shared'
import { RootStackParamList } from 'libs/shared/navigation/Stack'
import { ONBOARDING_VIEW_DATA } from '../constants'

import styles from './styles'

export type OnboardingProps = NativeStackScreenProps<RootStackParamList, 'Onboarding'>

const Onboarding = ({ navigation }: OnboardingProps) => {
  const pagerView = useRef<PagerView>(null)
  const numOfPagerView = useRef<number>(ONBOARDING_VIEW_DATA.length - 1)
  const [currentPage, setCurrentPage] = useState<number>(0)
  const handleMoveToNextScreen = useCallback(() => {
    if (currentPage === numOfPagerView.current) navigation.navigate('Login')
    setCurrentPage((prevState) =>
      prevState !== numOfPagerView.current ? prevState + 1 : prevState
    )
  }, [currentPage, navigation])
  const handlePageSelected = useCallback((e: PagerViewOnPageSelectedEvent) => {
    setCurrentPage(e.nativeEvent.position)
  }, [])
  const renderViewPagerChildren: React.JSX.Element[] = useMemo(
    () =>
      ONBOARDING_VIEW_DATA.map(({ id, img, title }: IOnboardingScreenView) => (
        <View key={id} style={styles.page} collapsable>
          <View style={styles.imgContainer}>
            <Image source={img} />
          </View>
          <Paragraph
            content={title}
            maxWidth={300}
            fontWeight="$2"
            color="$color.primary"
            textAlign="center"
            letterSpacing="$3"
            lineHeight="$6"
            fontSize="$4"
          />
        </View>
      )),
    []
  )
  const renderPagerIndex: React.JSX.Element[] = useMemo(() => {
    const listIndex: React.JSX.Element[] = []
    for (let i = 0; i <= numOfPagerView.current; i += 1) {
      listIndex.push(
        <View key={i} style={[styles.dot, currentPage === i ? styles.active : styles.normal]} />
      )
    }

    return listIndex
  }, [currentPage])

  useEffect(() => {
    pagerView.current?.setPage(currentPage)
  }, [currentPage])

  return (
    <Feature feat="onboarding">
      <SafeAreaView style={[containerStyles.expand, styles.container]}>
        <View style={styles.bgContainer} />
        <PagerView
          initialPage={currentPage}
          onPageSelected={handlePageSelected}
          ref={pagerView}
          style={containerStyles.expand}
        >
          {renderViewPagerChildren}
        </PagerView>
        <XStack
          marginBottom="$space.9"
          space="$space.2.5"
          alignItems="center"
          justifyContent="center"
        >
          {renderPagerIndex}
        </XStack>
        <Button
          title={numOfPagerView.current === currentPage ? 'finish' : 'next'}
          marginHorizontal="$space.6"
          onPress={handleMoveToNextScreen}
        />
      </SafeAreaView>
    </Feature>
  )
}

export default Onboarding
