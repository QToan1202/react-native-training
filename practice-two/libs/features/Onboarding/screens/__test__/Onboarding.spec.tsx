import { NativeStackNavigationProp } from '@react-navigation/native-stack'
import { fireEvent, screen } from '@testing-library/react-native'
import { RouteProp } from '@react-navigation/native'

import { render } from 'test-utils'
import { RootStackParamList } from '@navigation/Stack'

import Onboarding from '../index'

const navigation = {
  navigate: jest.fn(),
  goBack: jest.fn(),
} as unknown as NativeStackNavigationProp<RootStackParamList, 'Onboarding'>

const route = {
  name: 'Onboarding',
} as unknown as RouteProp<RootStackParamList, 'Onboarding'>

describe('Testing onboarding screen', () => {
  beforeEach(() => {
    render(<Onboarding navigation={navigation} route={route} testID="viewpager" />)
  })

  it('Render screen successfully', () => {
    const component = render(<Onboarding navigation={navigation} route={route} />).toJSON()

    expect(component).toMatchSnapshot()
  })

  it('Next button change to finish button at the end of view pager', async () => {
    const button = screen.getByText(/next/)

    fireEvent.press(button)
    fireEvent.press(button)

    const finishBtn = await screen.findByText(/finish/)
    expect(finishBtn).toBeOnTheScreen()

    fireEvent.press(finishBtn)
    expect(navigation.navigate).toBeCalledTimes(1)
  })
})
