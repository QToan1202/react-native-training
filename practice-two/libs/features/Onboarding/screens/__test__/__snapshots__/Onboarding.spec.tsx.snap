// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing onboarding screen Render screen successfully 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  route={
    {
      "name": "Onboarding",
    }
  }
  style={
    [
      {
        "flex": 1,
      },
      {
        "backgroundColor": "#fff",
        "paddingVertical": 20,
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "backgroundColor": "#33907C",
        "height": "40%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 750,
      }
    }
  />
  <RNCViewPager
    initialPage={0}
    layoutDirection="ltr"
    onMoveShouldSetResponderCapture={[Function]}
    onPageScroll={[Function]}
    onPageScrollStateChanged={[Function]}
    onPageSelected={[Function]}
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "rowGap": 42,
            },
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
          ]
        }
      >
        <View
          style={
            {
              "backgroundColor": "#fff",
              "borderRadius": 10,
              "paddingBottom": 30,
              "paddingTop": 60,
            }
          }
        >
          <Image
            source={
              {
                "testUri": "../../../assets/onboarding/onboarding-1.png",
              }
            }
          />
        </View>
        <Text
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 20,
              },
              {
                "color": "#33907C",
                "fontFamily": "Montserrat_500Medium",
                "letterSpacing": -0.5,
                "lineHeight": 30,
                "maxWidth": 300,
                "textAlign": "center",
              },
            ]
          }
        >
          Empowering Artisans, Farmers & Micro Business
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "rowGap": 42,
            },
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
          ]
        }
      >
        <View
          style={
            {
              "backgroundColor": "#fff",
              "borderRadius": 10,
              "paddingBottom": 30,
              "paddingTop": 60,
            }
          }
        >
          <Image
            source={
              {
                "testUri": "../../../assets/onboarding/onboarding-2.png",
              }
            }
          />
        </View>
        <Text
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 20,
              },
              {
                "color": "#33907C",
                "fontFamily": "Montserrat_500Medium",
                "letterSpacing": -0.5,
                "lineHeight": 30,
                "maxWidth": 300,
                "textAlign": "center",
              },
            ]
          }
        >
          Connecting NGOs, Social Enterprises with Communities
        </Text>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "rowGap": 42,
            },
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
          ]
        }
      >
        <View
          style={
            {
              "backgroundColor": "#fff",
              "borderRadius": 10,
              "paddingBottom": 30,
              "paddingTop": 60,
            }
          }
        >
          <Image
            source={
              {
                "testUri": "../../../assets/onboarding/onboarding-3.png",
              }
            }
          />
        </View>
        <Text
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 20,
              },
              {
                "color": "#33907C",
                "fontFamily": "Montserrat_500Medium",
                "letterSpacing": -0.5,
                "lineHeight": 30,
                "maxWidth": 300,
                "textAlign": "center",
              },
            ]
          }
        >
          Donate, Invest & Support infrastructure projects
        </Text>
      </View>
    </View>
  </RNCViewPager>
  <View
    style={
      {
        "alignItems": "center",
        "columnGap": 10,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 50,
      }
    }
  >
    <View
      style={
        [
          {
            "alignSelf": "flex-start",
            "borderRadius": 50,
            "height": 12,
            "width": 12,
          },
          {
            "backgroundColor": "#33907C",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "alignSelf": "flex-start",
            "borderRadius": 50,
            "height": 12,
            "width": 12,
          },
          {
            "backgroundColor": "#33907C",
            "opacity": 0.6,
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "alignSelf": "flex-start",
            "borderRadius": 50,
            "height": 12,
            "width": 12,
          },
          {
            "backgroundColor": "#33907C",
            "opacity": 0.6,
          },
        ]
      }
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#33907C",
        "borderColor": "#33907C",
        "borderRadius": 50,
        "borderWidth": 2,
        "marginHorizontal": 34,
        "opacity": 1,
        "paddingVertical": 13,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "columnGap": 6,
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
            },
            {
              "fontSize": 16,
            },
            [
              {
                "color": "#fff",
              },
              {
                "fontFamily": "Montserrat_500Medium",
                "textAlign": "center",
                "textTransform": "capitalize",
              },
              undefined,
            ],
          ]
        }
      >
        next
      </Text>
    </View>
  </View>
</RNCSafeAreaView>
`;
